kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -110,18 +117,57 @@ private CompletableFuture<KeyContext> makeKeyContext(@NonNull K kekId) {
                 }).toCompletableFuture();
     }
 
-    @NonNull
     @Override
+    @NonNull
     @SuppressWarnings("java:S2445")
-    public CompletionStage<Void> encrypt(@NonNull String topicName,
-                                         int partition,
-                                         @NonNull EncryptionScheme<K> encryptionScheme,
-                                         @NonNull List<? extends Record> records,
-                                         @NonNull Receiver receiver) {
-        if (records.isEmpty()) {
-            return CompletableFuture.completedFuture(null);
+    public CompletionStage<MemoryRecords> encrypt(@NonNull String topicName,
+                                                  int partition,
+                                                  @NonNull EncryptionScheme<K> encryptionScheme,
+                                                  @NonNull MemoryRecords records,
+                                                  @NonNull IntFunction<ByteBufferOutputStream> bufferAllocator) {
+        if (records.firstBatch() == null) {
+            return CompletableFuture.completedFuture(records);
+        }
+        MemoryRecordsBuilder builder = recordsBuilder(allocateBufferForEncode(records, bufferAllocator), records);
+        var encryptionRequests = recordStream(records).toList();
+        CompletionStage<Void> result;
+        if (encryptionRequests.isEmpty()) {
+            result = CompletableFuture.completedFuture(null);
```

Comment:
This might be redundant now that we check for a null firstBatch, not 100% sure, the tests showed that an empty memory records would have a null firstbatch and it fails when you try to calculate the size for buffer allocation.

Reply:
After checking with Tom there is a case where there can be a non-null batch with zero records in it, it's possible for a compacted topic to have all the records compacted out of a batch but the batch metadata needs to be preserved. We need to take care with this, for now we short-circuit respond if there are no records.